plugins/renepay/flow.h

```diff
@@ -0,0 +1,324 @@
+void chan_extra_relax_fraction(
+		struct chan_extra* ce,
+		double fraction);
+
+
+/* Returns either NULL, or an entry from the hash */
+struct chan_extra_half *get_chan_extra_half_by_scid(struct chan_extra_map *chan_extra_map,
+						    const struct short_channel_id scid,
+						    int dir);
+/* If the channel is not registered, then a new entry is created. scid must be
+ * present in the gossmap. */
+struct chan_extra_half *
+get_chan_extra_half_by_chan_verify(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		const struct gossmap_chan *chan,
+		int dir);
+
+/* Helper if we have a gossmap_chan */
+struct chan_extra_half *get_chan_extra_half_by_chan(const struct gossmap *gossmap,
+						    struct chan_extra_map *chan_extra_map,
+						    const struct gossmap_chan *chan,
+						    int dir);
+
+/* An actual partial flow. */
+struct flow {
+	struct gossmap_chan const **path;
+	/* The directions to traverse. */
+	int *dirs;
+	/* Amounts for this flow (fees mean this shrinks across path). */
+	struct amount_msat *amounts;
+	/* Probability of success (0-1) */
+	double success_prob;
+};
+
+/* Helper to access the half chan at flow index idx */
+const struct half_chan *flow_edge(const struct flow *flow, size_t idx);
+
+/* A big number, meaning "don't bother" (not infinite, since you may add) */
+#define FLOW_INF_COST 100000000.0
+
+/* Cost function to send @f msat through @c in direction @dir,
+ * given we already have a flow of prev_flow. */
+double flow_edge_cost(const struct gossmap *gossmap,
+		      const struct gossmap_chan *c, int dir,
+		      const struct amount_msat known_min,
+		      const struct amount_msat known_max,
+		      struct amount_msat prev_flow,
+		      struct amount_msat f,
+		      double mu,
+		      double basefee_penalty,
+		      double delay_riskfactor);
+
+/* Function to fill in amounts and success_prob for flow. */
+void flow_complete(struct flow *flow,
+		   const struct gossmap *gossmap,
+		   struct chan_extra_map *chan_extra_map,
+		   struct amount_msat delivered);
+
+/* Compute the prob. of success of a set of concurrent set of flows. */
+double flow_set_probability(
+		struct flow ** flows,
+		struct gossmap const*const gossmap,
+		struct chan_extra_map * chan_extra_map);
+
+// TODO(eduardo): we probably don't need this. Instead we should have payflow
+// input.
+/* Once flow is completed, this can remove it from the extra_map */
+void remove_completed_flow(const struct gossmap *gossmap,
+			   struct chan_extra_map *chan_extra_map,
+			   struct flow *flow);
+// TODO(eduardo): we probably don't need this. Instead we should have payflow
+// input.
+void remove_completed_flow_set(const struct gossmap *gossmap,
+			   struct chan_extra_map *chan_extra_map,
+			   struct flow **flows);
+
+struct amount_msat flow_set_fee(struct flow **flows);
+
+/*
+ * mu (μ) is used as follows in the cost function:
+ *
+ *     -log((c_e + 1 - f_e) / (c_e + 1)) + μ fee
```

Comment:
yes the paper suggests a weighted arithmetic mean between routing cost and uncertainty cost. However given all the issues that one can see from this I want to open the discussion about other means (like harmonic as deiscussed in #4771 ) or at least adding laplacian smoothing by adding some constant (c.f.: https://en.wikipedia.org/wiki/Additive_smoothing )

Reply:
I thought of adding prob. cost P and routing cost R using a fraction f (0<=f<=1):
```
cost = P * f + (1-f) R
```
so that when f=1 you get the entire costs is related to uncertainty, while with f=0 the costs is only fees. The good thing about this choice is that the interval to search is compact [0,1], and one can do binary search to look for the best value of f. 
Then we upgraded this to integer numbers by adopting the current parameter `mu` that lies in the interval `[0,MU_MAX]` and 
```
cost = P*mu + R*(MU_MAX-mu)
```
in essence is the same but with integers.

Allow me some time to study the discussion #4771.